web/TrackPlayer/Player.ts

```diff
@@ -0,0 +1,208 @@
+import { State } from '../../src/constants/State';
+import type { Track, Progress, PlaybackState } from '../../src/interfaces';
+import { SetupNotCalledError } from './SetupNotCalledError';
+
+export class Player {
+  protected hasInitialized: boolean = false;
+  protected element?: HTMLMediaElement;
+  protected player?: shaka.Player;
+  protected _current?: Track = undefined;
+  protected _playWhenReady: boolean = false;
+  protected _state: PlaybackState = { state: State.None };
+
+  // current getter/setter
+  public get current(): Track | undefined {
+    return this._current;
+  }
+  public set current(cur: Track | undefined) {
+    this._current = cur;
+  }
+
+  // state getter/setter
+  public get state(): PlaybackState {
+    return this._state;
+  }
+  public set state(newState: PlaybackState) {
+    this._state = newState;
+  }
+
+  // playWhenReady getter/setter
+  public get playWhenReady(): boolean {
+    return this._playWhenReady;
+  }
+  public set playWhenReady(pwr: boolean) {
+    this._playWhenReady = pwr;
+  }
+
+  async setupPlayer() {
+    // shaka only runs in a browser
+    if (typeof window === 'undefined') return;
+    if (this.hasInitialized === true) {
+      // TODO: double check the structure of this error message
+      throw { code: 'player_already_initialized', message: 'The player is not initialized. Call setupPlayer first.' };
+    }
+
+    // @ts-ignore
+    const shaka = await import('shaka-player/dist/shaka-player.ui');
+    // Install built-in polyfills to patch browser incompatibilities.
+    shaka.polyfill.installAll();
+    // Check to see if the browser supports the basic APIs Shaka needs.
+    if (!shaka.Player.isBrowserSupported()) {
+      // This browser does not have the minimum set of APIs we need.
+      this.state = {
+        state: State.Error,
+        error: {
+          code: 'not_supported',
+          message: 'Browser not supported.',
+        },
+      };
+      throw new Error('Browser not supported.');
+    }
+
+    // build dom element and attach shaka-player
+    this.element = document.createElement('audio');
+    this.element.setAttribute('id', 'react-native-track-player');
+    this.player = new shaka.Player(this.element);
+
+    // Listen for relevant events events.
+    this.player!.addEventListener('error', (error: any) => {
+      // Extract the shaka.util.Error object from the event.
+      this.onError(error.detail);
+    });
+    this.element.addEventListener('ended', () => this.onTrackEnded());
```

Comment:
nitpick: can't we just do this? Were you having binding issues? 

```suggestion
    this.element.addEventListener('ended',  this.onTrackEnded);
```

We probably can even generalize this, either updating the state directly in the closure here or having a general 

```ts
onStateUpdate(state: string) {
   switch (state) { ... }
}

this.element.addEventListener('playing', () => this.onStateUpdate('playing'))
```

Reply:
@dcvz yes, actually for some reason there are issues with binding. I'm not exactly sure why and I didn't dig into it a ton.

Regarding the modification to the `onStateUpdate` pattern. Which direction are you looking to go with this?

### Option 1

```ts
onStateUpdate(state: string) {
   switch (state) { 
     ...
     case 'ended':
        this.onTrackEnded();
        break;
     ...
   }
}
```

### Option 2

```ts
onStateUpdate(state: string) {
   switch (state) { 
     ...
     case 'ended':
        // directly add logic for the end state
        break;
     ...
   }
}
```

My sense is that you're going for Option 2? However, this will make extending state change behavior a little trickier potentially, as if you want to completely override a behavior you'll need to be aware of all the other events logic.

Let me know which direction you'd like to go and I'll send a commit.

Reply:
I was going more for option 2 -- should someone be able to extend this functionality? we expect most users to subscribe to state changes events right?

Reply:
regarding the binding, you may have to do `.bind(this, 'param')` or make these `const funName = () => {}`

Reply:
We can use `.bind` but AFAIK bind does basically the same thing right? From the docs:

> The bind() method of [Function](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function) instances creates a new function that, when called, calls this function with its this keyword set to the provided value, and a given sequence of arguments preceding any provided when the new function is called.

If I'm understanding that correctly I feel stylistically lambdas are more syntactically clear, but I might be missing some nuance. I'll update to bind.

Reply:
@jspizziri inso t have strong feelings then. Let’s leave it as is :)

Reply:
Also, regarding the extension issue. Users shouldn't be able to extend things. But we extend them internally in `TrackPlayerModule` and `PlaylistPlayer`. I'll update to the generic function so you can see the difference.

Reply:
As a note, the `onError` one isn't in there since it has a few more parameters that are dependent upon the error received. LMK your thoughts on that, feel like I could've done a better job.